File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        answer = ListNode(None, None)
        head = answer

        while 1:
            Min = float('inf')
            index = -1
            for i, current in enumerate(lists):
                if current is None:
                    continue
                
                if Min > current.val:
                    Min = current.val
                    index = i

                if answer.val == Min:
                    break

            if index == -1:
                break

            answer.next = ListNode(Min)
            lists[index] = lists[index].next
            answer = answer.next


        return head.next

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists.length == 0){
            return null;
        }
        int len = lists.length;

        ListNode answer = new ListNode(-1000000, null);
        ListNode answerHead = answer;

        int minIndex = 0;
        int minVal = 100000;
        while (!isNull(lists)){
            for (int i = 0; i < len; i++){
                ListNode node = lists[i];
                if (node == null){
                    continue;     
                }

                if (answer.val == node.val){
                    minVal = node.val;
                    minIndex = i;
                    break;
                }

                if (minVal > node.val){
                    minVal = node.val;
                    minIndex = i;
                }
            }
            lists[minIndex] = lists[minIndex].next;

            ListNode newNode = new ListNode(minVal);
            answer.next = newNode;
            answer = newNode;
            minVal = 100000;
        }

        return answerHead.next;
    }

    boolean isNull(ListNode[] lists){
        int len = lists.length;
        for (int i = 0; i < len; i++){
            if (lists[i] != null){
                return false;
            }
        }
        return true;
    }
}